File: app/routers/polls.py
"""Group availability poll routes."""

from __future__ import annotations

import logging
from datetime import datetime

from fastapi import APIRouter, HTTPException, Request
from fastapi.responses import HTMLResponse
from app.templates_env import get_templates
from pydantic import BaseModel, EmailStr
from sqlalchemy import select
from sqlalchemy.orm import selectinload

from app.dependencies import CalendarDep, CurrentUser, DbSession, EmailDep
from app.models.poll import AvailabilityPoll, PollResponse, PollSlot, PollStatus
from app.services.auth import generate_token

logger = logging.getLogger(__name__)
router = APIRouter(prefix="/polls", tags=["polls"])
templates = get_templates()


# ---------------------------------------------------------------------------
# Create poll (authenticated)
# ---------------------------------------------------------------------------


class SlotInput(BaseModel):
    start_at: datetime
    end_at: datetime


class PollCreate(BaseModel):
    title: str
    description: str | None = None
    slots: list[SlotInput]
    expires_at: datetime | None = None


@router.post("")
async def create_poll(data: PollCreate, user: CurrentUser, db: DbSession):
    poll = AvailabilityPoll(
        creator_id=user.id,
        title=data.title,
        description=data.description,
        expires_at=data.expires_at,
        share_token=generate_token(),
    )
    db.add(poll)
    await db.flush()

    for slot in data.slots:
        db.add(PollSlot(poll_id=poll.id, start_at=slot.start_at, end_at=slot.end_at))

    await db.commit()
    await db.refresh(poll)
    return {"poll_id": poll.id, "share_token": poll.share_token}


# ---------------------------------------------------------------------------
# View poll (public via share token)
# ---------------------------------------------------------------------------


@router.get("/{share_token}", response_class=HTMLResponse)
async def view_poll(share_token: str, request: Request, db: DbSession):
    result = await db.execute(
        select(AvailabilityPoll)
        .where(AvailabilityPoll.share_token == share_token)
        .options(
            selectinload(AvailabilityPoll.slots),
            selectinload(AvailabilityPoll.responses),
            selectinload(AvailabilityPoll.creator),
        )
    )
    poll = result.scalar_one_or_none()
    if poll is None:
        raise HTTPException(status_code=404, detail="Poll not found")

    # Tally votes per slot
    tally = {slot.id: {"yes": 0, "if_needed": 0, "no": 0} for slot in poll.slots}
    for resp in poll.responses:
        for slot_id_str, answer in resp.responses.items():
            slot_id = int(slot_id_str)
            if slot_id in tally and answer in tally[slot_id]:
                tally[slot_id][answer] += 1

    return templates.TemplateResponse(
        "pages/poll.html",
        {"request": request, "poll": poll, "tally": tally},
    )


# ---------------------------------------------------------------------------
# Submit response (public)
# ---------------------------------------------------------------------------


class PollResponseSubmit(BaseModel):
    participant_email: EmailStr
    participant_name: str
    responses: dict[str, str]  # {slot_id: "yes"|"if_needed"|"no"}


@router.post("/{share_token}/respond")
async def submit_response(share_token: str, data: PollResponseSubmit, db: DbSession):
    result = await db.execute(
        select(AvailabilityPoll).where(AvailabilityPoll.share_token == share_token)
    )
    poll = result.scalar_one_or_none()
    if poll is None or poll.status != PollStatus.open:
        raise HTTPException(status_code=404, detail="Poll not found or closed")

    # Check for existing response from this email
    existing = await db.execute(
        select(PollResponse).where(
            PollResponse.poll_id == poll.id,
            PollResponse.participant_email == data.participant_email,
        )
    )
    resp = existing.scalar_one_or_none()

    if resp:
        resp.responses = data.responses
        resp.participant_name = data.participant_name
    else:
        resp = PollResponse(
            poll_id=poll.id,
            participant_email=data.participant_email,
            participant_name=data.participant_name,
            responses=data.responses,
        )
        db.add(resp)

    await db.commit()
    return {"submitted": True}


# ---------------------------------------------------------------------------
# Confirm winning slot (authenticated creator only)
# ---------------------------------------------------------------------------


@router.post("/{share_token}/confirm/{slot_id}")
async def confirm_poll_slot(
    share_token: str,
    slot_id: int,
    user: CurrentUser,
    db: DbSession,
    email: EmailDep,
    calendar: CalendarDep,
    request: Request,
):
    result = await db.execute(
        select(AvailabilityPoll)
        .where(AvailabilityPoll.share_token == share_token)
        .options(selectinload(AvailabilityPoll.slots), selectinload(AvailabilityPoll.responses))
    )
    poll = result.scalar_one_or_none()
    if poll is None or poll.creator_id != user.id:
        raise HTTPException(status_code=403, detail="Not authorized")

    slot_result = await db.execute(select(PollSlot).where(PollSlot.id == slot_id, PollSlot.poll_id == poll.id))
    slot = slot_result.scalar_one_or_none()
    if slot is None:
        raise HTTPException(status_code=404, detail="Slot not found")

    poll.status = PollStatus.confirmed
    poll.confirmed_slot_id = slot.id
    await db.commit()

    # Notify all participants
    base_url = str(request.base_url).rstrip("/")
    for resp in poll.responses:
        try:
            await email.send_poll_confirmation(
                poll_id=poll.id,
                recipient_email=resp.participant_email,
                recipient_name=resp.participant_name,
                poll_title=poll.title,
                start=slot.start_at,
                end=slot.end_at,
                recipient_timezone="UTC",
            )
        except Exception as exc:
            logger.error("Poll confirmation email failed: %s", exc)

    return {"confirmed": True, "slot_start": slot.start_at.isoformat()}


# ---------------------------------------------------------------------------
# List user's polls
# ---------------------------------------------------------------------------


@router.get("")
async def list_polls(user: CurrentUser, db: DbSession):
    result = await db.execute(
        select(AvailabilityPoll)
        .where(AvailabilityPoll.creator_id == user.id)
        .order_by(AvailabilityPoll.created_at.desc())
        .options(selectinload(AvailabilityPoll.slots), selectinload(AvailabilityPoll.responses))
    )
    polls = result.scalars().all()
    return [
        {
            "id": p.id,
            "title": p.title,
            "status": p.status,
            "share_token": p.share_token,
            "expires_at": p.expires_at.isoformat() if p.expires_at else None,
            "response_count": len(p.responses),
            "slot_count": len(p.slots),
        }
        for p in polls
    ]

File: ts/src/routes/polls.ts
/**
 * Group availability poll routes.
 * Mirrors app/routers/polls.py.
 *
 * Routes:
 *   GET  /polls               → list current user's polls (auth required)
 *   POST /polls               → create a poll (auth required)
 *   GET  /polls/:shareToken   → view poll by share token (public)
 *   POST /polls/:shareToken/respond → submit availability response (public)
 *   POST /polls/:shareToken/confirm/:slotId → confirm winning slot (auth, creator only)
 */

import { Hono } from "hono";
import { zValidator } from "@hono/zod-validator";
import { z } from "zod";
import { and, eq } from "drizzle-orm";
import {
  getDb,
  availabilityPolls,
  pollSlots,
  pollResponses,
} from "@/db";
import { requireAuth } from "@/middleware/auth";
import { generateToken } from "@/auth/token-utils";
import { NoopEmailProvider } from "@/providers/noop";
import type { EmailProvider } from "@/providers/types";

type Env = { Variables: { userId: number } };
const polls = new Hono<Env>();

// Email provider injection (default noop; overridden in app setup)
let _emailProvider: EmailProvider = new NoopEmailProvider();
export function setPollEmailProvider(p: EmailProvider): void {
  _emailProvider = p;
}

function first<T>(arr: T[]): T | undefined {
  return arr[0];
}

// ---------------------------------------------------------------------------
// List user's polls
// ---------------------------------------------------------------------------

polls.get("/", requireAuth, async (c) => {
  const userId = c.get("userId");
  const db = getDb();

  const userPolls = await db
    .select()
    .from(availabilityPolls)
    .where(eq(availabilityPolls.creatorId, userId))
    .orderBy(availabilityPolls.createdAt);

  // For each poll, count slots and responses
  const result = await Promise.all(
    userPolls.map(async (p) => {
      const slots = await db
        .select({ id: pollSlots.id })
        .from(pollSlots)
        .where(eq(pollSlots.pollId, p.id));

      const responses = await db
        .select({ id: pollResponses.id })
        .from(pollResponses)
        .where(eq(pollResponses.pollId, p.id));

      return {
        id: p.id,
        title: p.title,
        status: p.status,
        shareToken: p.shareToken,
        expiresAt: p.expiresAt?.toISOString() ?? null,
        responseCount: responses.length,
        slotCount: slots.length,
      };
    }),
  );

  return c.json(result);
});

// ---------------------------------------------------------------------------
// Create poll
// ---------------------------------------------------------------------------

const slotInputSchema = z.object({
  startAt: z.string().datetime(),
  endAt: z.string().datetime(),
});

const pollCreateSchema = z.object({
  title: z.string().min(1),
  description: z.string().nullable().optional(),
  slots: z.array(slotInputSchema).min(1),
  expiresAt: z.string().datetime().nullable().optional(),
});

polls.post("/", requireAuth, zValidator("json", pollCreateSchema), async (c) => {
  const userId = c.get("userId");
  const data = c.req.valid("json");
  const db = getDb();

  const shareToken = generateToken();

  const [poll] = await db
    .insert(availabilityPolls)
    .values({
      creatorId: userId,
      title: data.title,
      description: data.description ?? null,
      expiresAt: data.expiresAt ? new Date(data.expiresAt) : null,
      shareToken,
    })
    .returning();

  if (!poll) {
    return c.json({ error: "Failed to create poll" }, 500);
  }

  if (data.slots.length > 0) {
    await db.insert(pollSlots).values(
      data.slots.map((s) => ({
        pollId: poll.id,
        startAt: new Date(s.startAt),
        endAt: new Date(s.endAt),
      })),
    );
  }

  return c.json({ pollId: poll.id, shareToken: poll.shareToken }, 201);
});

// ---------------------------------------------------------------------------
// View poll (public via share token)
// ---------------------------------------------------------------------------

polls.get("/:shareToken", async (c) => {
  const shareToken = c.req.param("shareToken");
  if (!shareToken) return c.json({ error: "Not found" }, 404);

  const db = getDb();

  const poll = first(
    await db
      .select()
      .from(availabilityPolls)
      .where(eq(availabilityPolls.shareToken, shareToken))
      .limit(1),
  );
  if (!poll) return c.json({ error: "Poll not found" }, 404);

  const slots = await db
    .select()
    .from(pollSlots)
    .where(eq(pollSlots.pollId, poll.id))
    .orderBy(pollSlots.startAt);

  const responses = await db
    .select()
    .from(pollResponses)
    .where(eq(pollResponses.pollId, poll.id));

  // Tally votes per slot
  const tally: Record<number, { yes: number; if_needed: number; no: number }> =
    {};
  for (const slot of slots) {
    tally[slot.id] = { yes: 0, if_needed: 0, no: 0 };
  }
  for (const resp of responses) {
    for (const [slotIdStr, answer] of Object.entries(resp.responses)) {
      const slotId = parseInt(slotIdStr, 10);
      const entry = tally[slotId];
      if (entry && (answer === "yes" || answer === "if_needed" || answer === "no")) {
        entry[answer]++;
      }
    }
  }

  return c.json({
    poll: {
      id: poll.id,
      title: poll.title,
      description: poll.description,
      status: poll.status,
      expiresAt: poll.expiresAt?.toISOString() ?? null,
      confirmedSlotId: poll.confirmedSlotId,
    },
    slots: slots.map((s) => ({
      id: s.id,
      startAt: s.startAt.toISOString(),
      endAt: s.endAt.toISOString(),
      tally: tally[s.id] ?? { yes: 0, if_needed: 0, no: 0 },
    })),
    responses: responses.map((r) => ({
      id: r.id,
      participantEmail: r.participantEmail,
      participantName: r.participantName,
      responses: r.responses,
    })),
  });
});

// ---------------------------------------------------------------------------
// Submit response (public)
// ---------------------------------------------------------------------------

const pollResponseSchema = z.object({
  participantEmail: z.string().email(),
  participantName: z.string().min(1),
  responses: z.record(z.enum(["yes", "if_needed", "no"])),
});

polls.post(
  "/:shareToken/respond",
  zValidator("json", pollResponseSchema),
  async (c) => {
    const shareToken = c.req.param("shareToken");
    if (!shareToken) return c.json({ error: "Not found" }, 404);

    const data = c.req.valid("json");
    const db = getDb();

    const poll = first(
      await db
        .select()
        .from(availabilityPolls)
        .where(
          and(
            eq(availabilityPolls.shareToken, shareToken),
            eq(availabilityPolls.status, "open"),
          ),
        )
        .limit(1),
    );
    if (!poll) return c.json({ error: "Poll not found or closed" }, 404);

    // Upsert by email
    const existing = first(
      await db
        .select()
        .from(pollResponses)
        .where(
          and(
            eq(pollResponses.pollId, poll.id),
            eq(pollResponses.participantEmail, data.participantEmail),
          ),
        )
        .limit(1),
    );

    if (existing) {
      await db
        .update(pollResponses)
        .set({
          responses: data.responses,
          participantName: data.participantName,
          updatedAt: new Date(),
        })
        .where(eq(pollResponses.id, existing.id));
    } else {
      await db.insert(pollResponses).values({
        pollId: poll.id,
        participantEmail: data.participantEmail,
        participantName: data.participantName,
        responses: data.responses,
      });
    }

    return c.json({ submitted: true });
  },
);

// ---------------------------------------------------------------------------
// Confirm winning slot (auth, creator only)
// ---------------------------------------------------------------------------

polls.post(
  "/:shareToken/confirm/:slotId",
  requireAuth,
  async (c) => {
    const shareToken = c.req.param("shareToken");
    const slotIdStr = c.req.param("slotId");
    if (!shareToken || !slotIdStr) return c.json({ error: "Not found" }, 404);

    const slotId = parseInt(slotIdStr, 10);
    const userId = c.get("userId");
    const db = getDb();

    const poll = first(
      await db
        .select()
        .from(availabilityPolls)
        .where(eq(availabilityPolls.shareToken, shareToken))
        .limit(1),
    );

    if (!poll || poll.creatorId !== userId) {
      return c.json({ error: "Not authorized" }, 403);
    }

    const slot = first(
      await db
        .select()
        .from(pollSlots)
        .where(
          and(eq(pollSlots.id, slotId), eq(pollSlots.pollId, poll.id)),
        )
        .limit(1),
    );
    if (!slot) return c.json({ error: "Slot not found" }, 404);

    // Update poll status
    await db
      .update(availabilityPolls)
      .set({
        status: "confirmed",
        confirmedSlotId: slot.id,
        updatedAt: new Date(),
      })
      .where(eq(availabilityPolls.id, poll.id));

    // Notify all participants
    const responses = await db
      .select()
      .from(pollResponses)
      .where(eq(pollResponses.pollId, poll.id));

    for (const resp of responses) {
      try {
        await _emailProvider.sendPollConfirmation({
          pollId: poll.id,
          recipientEmail: resp.participantEmail,
          recipientName: resp.participantName,
          pollTitle: poll.title,
          start: slot.startAt,
          end: slot.endAt,
          recipientTimezone: "UTC",
        });
      } catch (e) {
        console.error("Poll confirmation email failed:", e);
      }
    }

    return c.json({ confirmed: true, slotStart: slot.startAt.toISOString() });
  },
);

// ---------------------------------------------------------------------------
// Delete poll (auth, creator only)
// ---------------------------------------------------------------------------

polls.delete("/:shareToken", requireAuth, async (c) => {
  const shareToken = c.req.param("shareToken");
  if (!shareToken) return c.json({ error: "Not found" }, 404);

  const userId = c.get("userId");
  const db = getDb();

  const poll = first(
    await db
      .select()
      .from(availabilityPolls)
      .where(eq(availabilityPolls.shareToken, shareToken))
      .limit(1),
  );
  if (!poll || poll.creatorId !== userId) {
    return c.json({ error: "Not authorized" }, 403);
  }

  await db.delete(availabilityPolls).where(eq(availabilityPolls.id, poll.id));
  return c.json({ deleted: true });
});

export { polls as pollsRouter };
